### PR TITLE
feat(ui): Add "set as default" button to environment list

### PR DESCRIPTION
**Pull request by billyvg**
* Remove "Default Environment" settings from "Project General Settings"
* Change project update action creator to return promise

set_default_env

### CODE REVIEW

File: src/sentry/static/sentry/app/views/projectEnvironments.jsx [hunk 1]
```diff
@@ -119,6 +141,49 @@ const ProjectEnvironments = createReactClass({
...
+
+    // Update project details
+    update(this.api, {
+      ...this.props.params,
+      data,
+    }).then(
+      () => {
+        addSuccessMessage(
+          tct('Changed default environment to [environment]', {
+            environment: env.name,
```
Comment (lynnagara) [inline src/sentry/static/sentry/app/views/projectEnvironments.jsx]: env.displayName would be more consistent

**Comment by lynnagara** [comment]
We don't allow the user to set "" as a default environment right now, so it will be ignored

**Comment by lynnagara** [comment]
Any way to cut down that snapshot a bit? it's > 2000 lines now, quite hard to read

**Comment by lynnagara** [comment]
can we add tests for the cases where this field is 1. null and 2. an invalid string

File: src/sentry/static/sentry/app/views/projectEnvironments.jsx [hunk 2]
```diff
@@ -119,6 +141,49 @@ const ProjectEnvironments = createReactClass({
...
+      },
+      err => {
+        // Error occurred, revert project state
+        this.setState(state => ({
+          ...state,
+          project: oldProject,
+        }));
+        addErrorMessage(
+          tct('Unable to change default environment to [environment]', {
+            environment: env.name,
```
Comment (lynnagara) [inline src/sentry/static/sentry/app/views/projectEnvironments.jsx]: displayName

**Comment by billyvg** [comment]
@lynnagara what should we do in the the case of `""` and `null`? And what's considered an invalid string here, don't we pretty much accept anything as an env name?

**Comment by lynnagara** [comment]
I'm a bit conflicted about what to do with `""` and `null`. The main reason we ignored `""` before was that it was a text field before and we interpreted it as the absence of a value (there was no way to reset it back to `null` once it had been populated once). We could change that by modifying this line (https://github.com/getsentry/sentry/blob/master/src/sentry/static/sentry/app/views/projects/projectContext.jsx#L177) however my concern would be that it would change the default for the potentially large (not sure about exact #) number of people that have this value set already and that might be annoying in the short term at least

**Comment by lynnagara** [comment]
Re: invalid string - bad wording... I meant what happens if the field contains a value that is not an active environment in that project. e.g. Would it mark nothing as default, or a fallback value? Might be nice to have a test that specifies what happens in that case.

**Comment by billyvg** [comment]
@lynnagara Updated:

set_default_env

cc @saragilford 

Might be better to move invalid to last row?

**Comment by lynnagara** [comment]
Why is there a hide button only for "Dev"?

**Comment by lynnagara** [comment]
I wouldn't reorder the "" aka (no environment), just preserve the order it is returned from the API, which is what we do everywhere else it's displayed

**Review by lynnagara — APPROVED** [review]

**Pull request merged**